File: 98-validate-binary-search-tree/98-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

//Recursive traversal with valid range
// class Solution {
// public:
//     bool isValidBST(TreeNode* root) {
//      return validate(root,NULL,NULL);   
//     }
    
//     bool validate(TreeNode* root, TreeNode* low, TreeNode* high)
//     {
//         if(root==NULL)
//             return true;
        
//         if((low!=NULL && low->val >= root->val)||(high!=NULL && high->val <= root->val))
//             return false;
        
//         return validate(root->left,low,root) && validate(root->right,root,high);
//     }
// };


//Recursive inorder Traversal
class Solution {
public:
    TreeNode* prev;
    bool isValidBST(TreeNode* root) {
        prev = NULL;
        return inorder(root);
    }
    
    bool inorder(TreeNode* root)
    {
        if(root==NULL)
            return true;
        if(!inorder(root->left))
            return false;
        if(prev!=NULL && root->val <= prev->val)
            return false;
        prev = root;
        return inorder(root->right);
    }
};

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

//RECURSION APPROACH WITH A RANGE
// class Solution {
//     public boolean isValidBST(TreeNode root) {
//         return validBST(root,null,null);
        
//     }
    
//     boolean validBST(TreeNode root,Integer low, Integer high)
//     {
//         if(root==null)
//             return true;
        
//         if((low!=null && low > root.val) || (high!=null && high < root.val))
//             return false;
        
//         return validBST(root.left,low,root.val) && validBST(root.right,root.val,high);
//     } 
// }

//RECURSIVE APPROACH USING INORDER
// class Solution {
    
//     Integer prev = null;
    
//     public boolean isValidBST(TreeNode root) {
        
//         if(root==null)
//             return true;
        
//         if(!isValidBST(root.left))
//             return false;
        
//         if(prev!=null && root.val<=prev)
//             return false;
//         prev = root.val;
        
//         return isValidBST(root.right);
        
//     }
// }

//ITERATIVE SOLUTION USING INORDER
class Solution {
    public boolean isValidBST(TreeNode root) {
        if(root==null)
            return true;
        
        Stack<TreeNode> stack = new Stack<>();
        
        TreeNode prev = null;
        TreeNode curr=root;
        
        while(curr!=null || !stack.isEmpty())
        {
            while(curr!=null)
            {
                stack.push(curr);
                curr=curr.left;
            }
            curr = stack.pop();
            if(prev!=null && curr.val<=prev.val)
            {
                return false;                
            }
            prev = curr;
            curr=curr.right;
        }
        return true;
    }
}